Dev/dpopp/gpio interrupts

ddp5730 opened the pull request:
Implemented functionality to register custom IRQ Handler methods for GPIO pins.  Functionality mostly adapted from mbed_applications.

Resolved a few issues
- GPIO pins now properly initialized with direction when set up.
- GPIO pins now have a pull down resistor to keep pins from being in a floating state during most cases.  We may want functionality to modify this behavior at the user's request.

cbolles reviewed (changes requested):
Looks really good. One thing I would like to see if possible. Most interrupts generally will take in as a parameter the data being operated on (so in this case a pointer to a GPIO pin).

Ideally the function header for the interrupt would be something like

```
void myIRQ(IO::GPIO *pin);
```

This avoids the need for global variables. It's honestly a flaw in the HAL code more then anything. But having the value passed in is a practice common elsewhere especially in the Linux kernel. For our cases we can probably implement this by having another array in the GPIO code that stores the pointer to the GPIO that is associated with the corresponding function pointer.
Something like.

```
void (*INTERRUPT_HANDLERS[16])(*IO::GPIO) = {0};
IO::GPIO* INTERRUPT_PARAMS[16] = {0};
```
 So that you can then call the interrupt like

```
INTERRUPT_HANDLERS[count](INTERRUPT_PARAMS[count]);
```

Other than that, a couple of stylistic changes, and it looks really good!

ddp5730 commented on include/EVT/io/GPIO.hpp:
```diff
@@ -83,6 +94,13 @@ class GPIO
+     * condition.
+     * @param edge The edge trigger event to trigger the interrupt
+     */
+    virtual void registerIrq(TriggerEdge edge, EVT::core::types::void_function_ptr_t irqHandler) = 0;
```
I agree about the quick capitalization thing.  

As for the typing, I actually do think you could make an argument that a function pointer is a type.  You use the cpp syntax to declare a type to create the void_function_pointer_t type.
However, I could see it either way.  
The other version looks like this `virtual void registerIRQ(TriggerEdge edge, const void (*irqHandler)()) = 0;`  I'll make the change, I just feel like it's less readable.

ddp5730 commented on include/EVT/utils/types/VoidFunctions.h:
```diff
@@ -0,0 +1,15 @@
+/**
```
Good point, I just missed that here.  I removed the file anyways.

ddp5730 commented on samples/gpio_interrupt/main.cpp:
```diff
@@ -0,0 +1,45 @@
+
+namespace IO = EVT::core::IO;
+
+constexpr int BAUD_RATE = 115200;
```
Good call.  How do you feel about us having some common variables/constants file that holds the common BAUD_RATE and other variables across our projects.  That way we can easily enforce a standard within a project?

cbolles replied:
I'm on the fence, on one hand I love that it would create unity, on the other hand I like transparency in the samples since it makes it easier for people just starting out on the team (don't have to do a bunch of clicking around to find out why things are the way that they are). I could really be persuaded either way.

ddp5730 replied:
I see your point and I agree.

ddp5730 commented:
> Looks really good. One thing I would like to see if possible. Most interrupts generally will take in as a parameter the data being operated on (so in this case a pointer to a GPIO pin).
> 
> Ideally the function header for the interrupt would be something like
> 
> ```
> void myIRQ(IO::GPIO *pin);
> ```
> 
> This avoids the need for global variables. It's honestly a flaw in the HAL code more then anything. But having the value passed in is a practice common elsewhere especially in the Linux kernel. For our cases we can probably implement this by having another array in the GPIO code that stores the pointer to the GPIO that is associated with the corresponding function pointer.
> Something like.
> 
> ```
> void (*INTERRUPT_HANDLERS[16])(*IO::GPIO) = {0};
> IO::GPIO* INTERRUPT_PARAMS[16] = {0};
> ```
> 
> So that you can then call the interrupt like
> 
> ```
> INTERRUPT_HANDLERS[count](INTERRUPT_PARAMS[count]);
> ```
> 
> Other than that, a couple of stylistic changes, and it looks really good!

I really like your suggestion for having a resource pointer.  Implemented and I've addressed all PR comments and pushed updates.  Please review again when you have time.

cbolles reviewed (approved):
Look good!
